src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -274,60 +274,80 @@ bool AbstractLQPNode::has_matching_unique_constraint(const ExpressionUnorderedSe
 }
 
 std::vector<FunctionalDependency> AbstractLQPNode::functional_dependencies() const {
-  // Gather input FDs from all input nodes
-  auto fds_left = std::vector<FunctionalDependency>();
-  auto fds_right = std::vector<FunctionalDependency>();
-  if (left_input()) {
-    fds_left = left_input()->functional_dependencies();
-  }
-  if (right_input()) {
-    fds_right = right_input()->functional_dependencies();
+  // (1) Gather FDs from previous nodes
+  auto fds_in = pass_functional_dependencies();
+  if constexpr (HYRISE_DEBUG) {
+    auto distinct_fds = std::vector<FunctionalDependency>();
+    for (const auto& fd : fds_in) {
+      Assert(std::none_of(distinct_fds.cbegin(), distinct_fds.cend(),
+                          [&fd](const auto& distinct_fd) { return fd == distinct_fd; }),
+             "Did not expect LQP nodes to pass duplicate FDs.");
+      distinct_fds.push_back(fd);
+    }
   }
 
-  if (HYRISE_DEBUG && !fds_right.empty()) {
-    for (const auto& fd_right : fds_right) {
-      const bool duplicate = std::any_of(fds_left.begin(), fds_left.end(),
-                                         [&fd_right](const auto& fd_left) { return (fd_left == fd_right); });
-      Assert(!duplicate,
-             "Unexpected duplicate functional dependency found in " + this->description(DescriptionMode::Short));
+  // (2) Generate FDs from node's unique constraints
+  const auto& output_expressions = this->output_expressions();
+  const auto unique_constraints = this->unique_constraints();
+  for (const auto& unique_constraint : *unique_constraints) {
+    auto determinants = unique_constraint.expressions;
+    auto dependents = ExpressionUnorderedSet();
+    for (const auto& output_expression : output_expressions) {
+      if (!determinants.contains(output_expression)) {
+        dependents.insert(output_expression);
+      }
+    }
+
+    // Add the FD to the output if not already existing
```

Comment:
Maybe I haven't fully understood the concept yet. In which case would the FD already exist? Wouldn't that mean that some input node passed an FD that is based on a unique constraint?

Reply:
This block is due to the special handling of joins. 
In JoinNode, we prematurely create FDs from unique constraints because we might lose some in upper nodes.
To simplify the implementation, JoinNode creates FDs from all unique constraints. 

However, not all unique constraints get lost. Therefore, we might end up with FD duplicates here.

Reply:
Hmm, but doesn't that violate the "non-derivable/non-trivial FD" contract discussed above? Having a method that does not forward unique constraints *unless there was a join somewhere below that node* seems hard to deal with.